engine/pkg/dm/message_agent.go

```diff
@@ -0,0 +1,407 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dm
+
+import (
+	"context"
+	"encoding/hex"
+	"encoding/json"
+	"path"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tiflow/dm/pkg/log"
+	"github.com/pingcap/tiflow/engine/lib"
+	"github.com/pingcap/tiflow/engine/pkg/p2p"
+	"github.com/pingcap/tiflow/pkg/workerpool"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+)
+
+var (
+	defaultMessageTimeOut  = time.Second * 10
+	defaultRequestTimeOut  = time.Second * 30
+	defaultResponseTimeOut = time.Second * 10
+	defaultHandlerTimeOut  = time.Second * 30
+)
+
+// generateTopic generate dm message topic with hex encoding.
+func generateTopic(senderID string, receiverID string) string {
+	hexKeys := []string{"DM"}
+	hexKeys = append(hexKeys, hex.EncodeToString([]byte(senderID)))
+	hexKeys = append(hexKeys, hex.EncodeToString([]byte(receiverID)))
+	ret := path.Join(hexKeys...)
+	return ret
+}
+
+// extractTopic extract dm message topic with hex decoding.
+// TODO: handle error.
+func extractTopic(topic string) (string, string) {
+	v := strings.Split(strings.TrimPrefix(topic, "DM"), "/")
+	// nolint:errcheck
+	senderID, _ := hex.DecodeString(v[1])
+	// nolint:errcheck
+	receiverID, _ := hex.DecodeString(v[2])
```

Comment:
maybe receiverID can't be empty?

```golang
func FuzzTopic(f *testing.F) {
	f.Add("id1", "id2")
	f.Fuzz(func(t *testing.T, id1, id2 string) {
		topic := generateTopic(id1, id2)
		gotID1, gotID2 := extractTopic(topic)
		require.Equal(t, id1, gotID1)
		require.Equal(t, id2, gotID2)
	})
}
```

```
C:\Users\me\go\src\github.com\pingcap\tiflow\engine\pkg\dm>SET CGO_ENABLED=0& go test -fuzz FuzzTopic
fuzz: elapsed: 0s, gathering baseline coverage: 0/1 completed
fuzz: elapsed: 0s, gathering baseline coverage: 1/1 completed, now fuzzing with 8 workers
fuzz: elapsed: 1s, execs: 124 (120/sec), new interesting: 1 (total: 2)
--- FAIL: FuzzTopic (1.04s)
    --- FAIL: FuzzTopic (0.00s)
        testing.go:1349: panic: runtime error: index out of range [2] with length 2
            goroutine 275 [running]:
            runtime/debug.Stack()
                C:/Program Files/Go/src/runtime/debug/stack.go:24 +0x90
            testing.tRunner.func1()
                C:/Program Files/Go/src/testing/testing.go:1349 +0x1f2
            panic({0x3398ee0, 0xc000cc8048})
                C:/Program Files/Go/src/runtime/panic.go:838 +0x207
            github.com/pingcap/tiflow/engine/pkg/dm.extractTopic({0xc000e20f96, 0x5})
                C:/Users/TianYuan/go/src/github.com/pingcap/tiflow/engine/pkg/dm/message_agent.go:58 +0x22f
            github.com/pingcap/tiflow/engine/pkg/dm.FuzzTopic.func1(0x5?, {0xc000e20e11, 0x1}, {0x0, 0x0})
                C:/Users/TianYuan/go/src/github.com/pingcap/tiflow/engine/pkg/dm/message_agent_test.go:317 +0x65
            reflect.Value.call({0x30c8ba0?, 0x36f11b0?, 0x60?}, {0x3590202, 0x4}, {0xc000092ae0, 0x3, 0x3ca410?})
                C:/Program Files/Go/src/reflect/value.go:556 +0x845
            reflect.Value.Call({0x30c8ba0?, 0x36f11b0?, 0x514?}, {0xc000092ae0, 0x3, 0x4})
                C:/Program Files/Go/src/reflect/value.go:339 +0xbf
            testing.(*F).Fuzz.func1.1(0x0?)
                C:/Program Files/Go/src/testing/fuzz.go:337 +0x231
            testing.tRunner(0xc000506820, 0xc000f0f830)
                C:/Program Files/Go/src/testing/testing.go:1439 +0x102
            created by testing.(*F).Fuzz.func1
                C:/Program Files/Go/src/testing/fuzz.go:324 +0x5b8


    Failing input written to testdata\fuzz\FuzzTopic\85649d45641911d0afeadb6dc0dd894d90e5f634730e84468836ba0446fc6f38
    To re-run:
    go test -run=FuzzTopic/85649d45641911d0afeadb6dc0dd894d90e5f634730e84468836ba0446fc6f38
FAIL
exit status 1
FAIL    github.com/pingcap/tiflow/engine/pkg/dm 4.015s
```

Reply:
I add the TODO comment: "handle error", now we use masterID/taskID as senderID/receiverID, it should not be empty.